Sources/Hub/Downloader.swift

```diff
@@ -0,0 +1,122 @@
+//
+//  Downloader.swift
+//
+//  Adapted from https://github.com/huggingface/swift-coreml-diffusers/blob/d041577b9f5e201baa3465bc60bc5d0a1cf7ed7f/Diffusion/Common/Downloader.swift
+//  Created by Pedro Cuenca on December 2022.
+//  See LICENSE at https://github.com/huggingface/swift-coreml-diffusers/LICENSE
+//
+
+import Foundation
+import Combine
+
+class Downloader: NSObject, ObservableObject {
+    private(set) var destination: URL
+    
+    enum DownloadState {
+        case notStarted
+        case downloading(Double)
+        case completed(URL)
+        case failed(Error)
+    }
+    
+    enum DownloadError: Error {
+        case invalidDownloadLocation
+        case unexpectedError
+    }
+    
+    private(set) lazy var downloadState: CurrentValueSubject<DownloadState, Never> = CurrentValueSubject(.notStarted)
+    private var stateSubscriber: Cancellable?
+    
+    private var urlSession: URLSession? = nil
+    
+    init(from url: URL, to destination: URL, using authToken: String? = nil) {
```

Comment:
Downloader accepts an authentication token, but we don't expose this in the `HubApi` yet. It should just be another ivar.